dhall/src/builtins.rs

```diff
@@ -401,6 +409,62 @@ fn apply_builtin(b: Builtin, args: Vec<Nir>, env: NzEnv) -> NirKind {
             }
             _ => Ret::DoneAsIs,
         },
+        (Builtin::TextReplace, [needle, replacement, haystack]) => {
+            match (&*needle.kind(), &*replacement.kind(), &*haystack.kind()) {
+                (TextLit(n_lit), TextLit(r_lit), TextLit(h_lit)) => {
+                    // When the needle is empty we just return the haystack untouched
+                    if n_lit.is_empty() {
+                        Ret::Nir(haystack.clone())
+                    // The needle needs to be fully evaluated as Text
+                    // otherwise no progress can be made
+                    } else if let Some(n) = n_lit.as_text() {
+                        // Fast case when replacement and haystack are fully evaluated
```

Comment:
I don't see why it makes sense to replace if the haystack is not fully evaluated. For example, in `\(x : Text) -> Text/replace "aa" "b" "a${x}"`, we can't know if replacing will occur. So I would completely forbid replacing if the haystack is not in normal form. Is this what the standard requires?

Reply:
Oh ok, that seems to be what you suggest in https://github.com/dhall-lang/dhall-lang/pull/1084